diff --git a/src/event/ngx_event_openssl.c b/src/event/ngx_event_openssl.c
index 91b415ca..f876e9c9 100644
--- a/src/event/ngx_event_openssl.c
+++ b/src/event/ngx_event_openssl.c
@@ -1588,6 +1588,48 @@ ngx_ssl_set_session(ngx_connection_t *c, ngx_ssl_session_t *session)
     return NGX_OK;
 }
 
+int
+ngx_ssl_client_hello_cb(SSL *s, int *al, void *arg) {
+    ngx_connection_t  *c = arg;
+
+    if (c == NULL) {
+        return 1;
+    }
+
+    if (c->ssl == NULL) {
+        return 1;
+    }
+
+    c->ssl->fp_extensions.data = NULL;
+    c->ssl->fp_extensions.len = 0;
+    SSL_client_hello_get_extension_order(c->ssl->connection, NULL, &c->ssl->fp_extensions.len);
+    if (c->ssl->fp_extensions.len) {
+        c->ssl->fp_extensions.data = ngx_pnalloc(c->pool, c->ssl->fp_extensions.len*2);
+        SSL_client_hello_get_extension_order(c->ssl->connection, (uint16_t*)c->ssl->fp_extensions.data, &c->ssl->fp_extensions.len);
+    }
+
+    return 1;
+}
+
+void
+ngx_ssl_client_hello_ext(ngx_connection_t *c) {
+    ngx_ssl_connection_t *ssl = c->ssl;
+
+    if (ssl == NULL) {
+        return;
+    }
+
+    ssl->fp_ciphers.data = NULL;
+    ssl->fp_ciphers.len = SSL_ctrl(ssl->connection, SSL_CTRL_GET_RAW_CIPHERLIST, 0, &ssl->fp_ciphers.data);
+
+    ssl->fp_groups.data = NULL;
+    ssl->fp_groups.len = SSL_ctrl(ssl->connection, SSL_CTRL_GET_IANA_GROUPS, 0, &ssl->fp_groups.data);
+
+    ssl->fp_points.data = NULL;
+    ssl->fp_points.len = SSL_ctrl(ssl->connection, SSL_CTRL_GET_EC_POINT_FORMATS, 0, &ssl->fp_points.data);
+
+    return;
+}
 
 ngx_int_t
 ngx_ssl_handshake(ngx_connection_t *c)
@@ -1603,8 +1645,12 @@ ngx_ssl_handshake(ngx_connection_t *c)
 
     ngx_ssl_clear_error(c->log);
 
+    SSL_CTX_set_client_hello_cb(c->ssl->session_ctx, ngx_ssl_client_hello_cb, c);
+
     n = SSL_do_handshake(c->ssl->connection);
 
+    ngx_ssl_client_hello_ext(c);
+
     ngx_log_debug1(NGX_LOG_DEBUG_EVENT, c->log, 0, "SSL_do_handshake: %d", n);
 
     if (n == 1) {
diff --git a/src/event/ngx_event_openssl.h b/src/event/ngx_event_openssl.h
index 61da0c5d..1317c416 100644
--- a/src/event/ngx_event_openssl.h
+++ b/src/event/ngx_event_openssl.h
@@ -99,6 +99,11 @@ struct ngx_ssl_connection_s {
     unsigned                    in_early:1;
     unsigned                    early_preread:1;
     unsigned                    write_blocked:1;
+
+    ngx_str_t                   fp_ciphers;
+    ngx_str_t                   fp_extensions;
+    ngx_str_t                   fp_groups;
+    ngx_str_t                   fp_points;
 };
 
 
diff --git a/src/http/v2/ngx_http_v2.c b/src/http/v2/ngx_http_v2.c
index 8b0fc53b..1eb377fe 100644
--- a/src/http/v2/ngx_http_v2.c
+++ b/src/http/v2/ngx_http_v2.c
@@ -325,6 +325,14 @@ ngx_http_v2_init(ngx_event_t *rev)
 
     c->idle = 1;
 
+    h2c->fp_fingerprinted = 0;
+    h2c->fp_settings.data = ngx_pnalloc(c->pool, 30);
+    h2c->fp_settings.len = 0;
+    h2c->fp_priorities.data = ngx_pnalloc(c->pool, 32);
+    h2c->fp_priorities.len = 0;
+    h2c->fp_pseudoheaders.data = ngx_pnalloc(c->pool, 32);
+    h2c->fp_pseudoheaders.len = 0;
+
     ngx_http_v2_read_handler(rev);
 }
 
@@ -1204,6 +1212,14 @@ ngx_http_v2_state_headers(ngx_http_v2_connection_t *h2c, u_char *pos,
         }
     }
 
+    if (!h2c->fp_fingerprinted && h2c->fp_priorities.len < 32) {
+        h2c->fp_priorities.data[h2c->fp_priorities.len] = (uint8_t)stream->node->id;
+        h2c->fp_priorities.data[h2c->fp_priorities.len+1] = (uint8_t)excl;
+        h2c->fp_priorities.data[h2c->fp_priorities.len+2] = (uint8_t)depend;
+        h2c->fp_priorities.data[h2c->fp_priorities.len+3] = (uint8_t)(weight-1);
+        h2c->fp_priorities.len += 4;
+    }
+
     return ngx_http_v2_state_header_block(h2c, pos, end);
 
 rst_stream:
@@ -1625,6 +1641,13 @@ ngx_http_v2_state_process_header(ngx_http_v2_connection_t *h2c, u_char *pos,
         goto error;
     }
 
+    if (!h2c->fp_fingerprinted && h2c->fp_pseudoheaders.len < 32) {
+        if (header->name.data[0] == ':' && header->name.len > 1)
+            h2c->fp_pseudoheaders.data[h2c->fp_pseudoheaders.len++] = header->name.data[1];
+        else
+            h2c->fp_pseudoheaders.data[h2c->fp_pseudoheaders.len++] = header->name.data[0];
+    }
+
     if (header->name.data[0] == ':') {
         rc = ngx_http_v2_pseudo_header(r, header);
 
@@ -2042,6 +2065,12 @@ ngx_http_v2_state_settings_params(ngx_http_v2_connection_t *h2c, u_char *pos,
         ngx_log_debug2(NGX_LOG_DEBUG_HTTP, h2c->connection->log, 0,
                        "http2 setting %ui:%ui", id, value);
 
+        if (!h2c->fp_fingerprinted && h2c->fp_settings.len < 30) {
+            h2c->fp_settings.data[h2c->fp_settings.len] = (uint8_t)id;
+            *(uint32_t*)(h2c->fp_settings.data + h2c->fp_settings.len + 1)  = (uint32_t)value;
+            h2c->fp_settings.len += 5;
+        }
+
         switch (id) {
 
         case NGX_HTTP_V2_INIT_WINDOW_SIZE_SETTING:
@@ -2344,6 +2373,7 @@ ngx_http_v2_state_window_update(ngx_http_v2_connection_t *h2c, u_char *pos,
         }
     }
 
+    h2c->fp_windowupdate = window;
     return ngx_http_v2_state_complete(h2c, pos, end);
 }
 
diff --git a/src/http/v2/ngx_http_v2.h b/src/http/v2/ngx_http_v2.h
index 59ddf54e..fd353631 100644
--- a/src/http/v2/ngx_http_v2.h
+++ b/src/http/v2/ngx_http_v2.h
@@ -163,6 +163,12 @@ struct ngx_http_v2_connection_s {
     unsigned                         blocked:1;
     unsigned                         goaway:1;
     unsigned                         push_disabled:1;
+
+    unsigned                         fp_fingerprinted:1;
+    ngx_str_t                        fp_settings;
+    ngx_uint_t                       fp_windowupdate;
+    ngx_str_t                        fp_priorities;
+    ngx_str_t                        fp_pseudoheaders;
 };
 
 
