diff --git a/include/openssl/ssl.h.in b/include/openssl/ssl.h.in
index 9ff923b..ba93c0d 100644
--- a/include/openssl/ssl.h.in
+++ b/include/openssl/ssl.h.in
@@ -1931,6 +1931,8 @@ size_t SSL_client_hello_get0_ciphers(SSL *s, const unsigned char **out);
 size_t SSL_client_hello_get0_compression_methods(SSL *s,
     const unsigned char **out);
 int SSL_client_hello_get1_extensions_present(SSL *s, int **out, size_t *outlen);
+size_t SSL_client_hello_get_ja_data(SSL *s, unsigned char *data,
+                                     size_t data_size);
 int SSL_client_hello_get_extension_order(SSL *s, uint16_t *exts,
     size_t *num_exts);
 int SSL_client_hello_get0_ext(SSL *s, unsigned int type,
diff --git a/include/openssl/tls1.h b/include/openssl/tls1.h
index 34d3768..019ace5 100644
--- a/include/openssl/tls1.h
+++ b/include/openssl/tls1.h
@@ -141,6 +141,13 @@ extern "C" {
 /* ExtensionType value from RFC7627 */
 #define TLSEXT_TYPE_extended_master_secret 23
 
+/* ExtensionType value from RFC6961 */
+# define TLSEXT_TYPE_status_request_v2           17
+/* ExtensionType value from RFC8449 */
+# define TLSEXT_TYPE_record_size_limit           28
+/* ExtensionType value from RFC7639 */
+# define TLSEXT_TYPE_application_settings        17513
+
 /* ExtensionType value from RFC8879 */
 #define TLSEXT_TYPE_compress_certificate 27
 
diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index 4016fde..99e471a 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -6869,6 +6869,175 @@ err:
     return 0;
 }
 
+size_t SSL_client_hello_get_ja_data(SSL *s, unsigned char *data,
+        size_t data_size)
+{
+    RAW_EXTENSION *ext;
+    const unsigned char *ext_data = NULL;
+    PACKET *groups = NULL, *formats = NULL,
+            *sigalgs = NULL, *supvers = NULL,
+            *alpn = NULL;
+    size_t num = 0, i;
+    uint16_t version = 0, n;
+    unsigned char *ptr = data,
+                  *end = data + data_size;
+    const SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(s);
+
+    if (sc == NULL)
+        return 0;
+
+    if (ossl_unlikely(sc->clienthello == NULL || data == NULL
+                /** just checking that we have at least this */
+                || data_size < 128))
+        return 0;
+
+    /* version */
+    *(uint16_t *) ptr = (uint16_t) sc->clienthello->legacy_version;
+    ptr += sizeof(uint16_t);
+
+    /* ciphers */
+    num = PACKET_remaining(&sc->clienthello->ciphersuites);
+    if (ossl_unlikely(ptr + sizeof(uint16_t) + num > end))
+        return 0;
+    *(uint16_t *) ptr = (uint16_t) num;
+    ptr += sizeof(uint16_t);
+    if (ossl_likely(num > 0)) {
+        memcpy(ptr, PACKET_data(&sc->clienthello->ciphersuites), num);
+        ptr += num;
+    }
+
+    /* extensions */
+    num = 0;
+    for (i = 0; i < sc->clienthello->pre_proc_exts_len; i++) {
+        ext = sc->clienthello->pre_proc_exts + i;
+        if (ext->present)
+            num++;
+    }
+
+    num = num * 2;
+    if (ossl_unlikely((ptr + num + sizeof(uint16_t)) > end))
+        return 0;
+
+    *(uint16_t*) ptr = (uint16_t) num;
+    ptr += sizeof(uint16_t);
+
+    for (i = 0; i < sc->clienthello->pre_proc_exts_len; i++) {
+        ext = sc->clienthello->pre_proc_exts + i;
+        if (ext->present) {
+            if (ext->received_order >= num)
+                break;
+            if (ossl_likely(ext->received_order < data_size)) {
+                ((uint16_t*)(ptr))[ext->received_order] = (uint16_t) ext->type;
+            }
+            switch (ext->type) {
+            case TLSEXT_TYPE_supported_groups:
+                groups = &ext->data;
+                break;
+            case TLSEXT_TYPE_ec_point_formats:
+                formats = &ext->data;
+                break;
+            case TLSEXT_TYPE_signature_algorithms:
+                sigalgs = &ext->data;
+                break;
+            case TLSEXT_TYPE_supported_versions:
+                supvers = &ext->data;
+                break;
+            case TLSEXT_TYPE_application_layer_protocol_negotiation:
+                alpn = &ext->data;
+                break;
+            }
+        }
+    }
+
+    ptr += num;
+
+    /* groups */
+    if (groups
+        && (num = PACKET_remaining(groups)) > 0)
+    {
+        if (ossl_unlikely((ptr + num + sizeof(uint16_t)) > end))
+            return 0;
+        memcpy(ptr, PACKET_data(groups), num);
+        *(uint16_t*) ptr = (uint16_t) num;
+        ptr += num;
+    } else {
+        if (ossl_unlikely(ptr + sizeof(uint16_t) > end))
+            return 0;
+        *(uint16_t *) ptr = (uint16_t) 0;
+        ptr += sizeof(uint16_t);
+    }
+
+    /* formats */
+    if (formats
+        && (num = PACKET_remaining(formats)) > 0)
+    {
+        if (ossl_unlikely((ptr + num + sizeof(uint8_t)) > end))
+            return 0;
+        memcpy(ptr, PACKET_data(formats), num);
+        *(uint8_t *) ptr = (uint8_t) num;
+        ptr += num;
+    } else {
+        if (ossl_unlikely(ptr + sizeof(uint8_t) > end))
+            return 0;
+        *(uint8_t *) ptr = (uint8_t) 0;
+        ptr += sizeof(uint8_t);
+    }
+
+    /* supported version */
+    if (supvers
+        && (num = PACKET_remaining(supvers)) >= 3)
+    {
+        ext_data = PACKET_data(supvers);
+        for (i = 1; i + 1 < num && i <= ext_data[0]; i += 2) {
+            n = ((uint16_t) ext_data[i] << 8) | (uint16_t) ext_data[i + 1];
+            if (((n & 0x0f0f) != 0x0a0a || (n & 0xff) != (n >> 8))
+                && n > version)
+            {
+                version = n;
+            }
+        }
+    }
+
+    if (ossl_unlikely(ptr + sizeof(uint16_t) > end))
+        return 0;
+    *(uint16_t *) ptr = version;
+    ptr += sizeof(uint16_t);
+
+    /* signature algorithms */
+    if (sigalgs
+        && (num = PACKET_remaining(sigalgs)) > 0)
+    {
+        if (ossl_unlikely(ptr + num > end))
+            return 0;
+        memcpy(ptr, PACKET_data(sigalgs), num);
+        *(uint16_t *) ptr = (uint16_t) num;
+        ptr += num;
+    } else {
+        if (ossl_unlikely(ptr + sizeof(uint16_t) > end))
+            return 0;
+        *(uint16_t *) ptr = (uint16_t) 0;
+        ptr += sizeof(uint16_t);
+    }
+
+    /* alpn */
+    if (alpn
+        && (num = PACKET_remaining(alpn)) > 0)
+    {
+        if (ossl_unlikely(ptr + num > end))
+            return 0;
+        memcpy(ptr, PACKET_data(alpn), num);
+        *(uint16_t *) ptr = (uint16_t) num;
+        ptr += num;
+    } else {
+        if (ossl_unlikely(ptr + sizeof(uint16_t) > end))
+            return 0;
+        *(uint16_t *) ptr = (uint16_t) 0;
+        ptr += sizeof(uint16_t);
+    }
+
+    return ptr - data;
+}
+
 int SSL_client_hello_get_extension_order(SSL *s, uint16_t *exts, size_t *num_exts)
 {
     RAW_EXTENSION *ext;
diff --git a/ssl/ssl_local.h b/ssl/ssl_local.h
index 2588ef5..4806882 100644
--- a/ssl/ssl_local.h
+++ b/ssl/ssl_local.h
@@ -686,6 +686,9 @@ typedef enum tlsext_index_en {
     TLSEXT_IDX_compress_certificate,
     TLSEXT_IDX_early_data,
     TLSEXT_IDX_certificate_authorities,
+    TLSEXT_IDX_status_request_v2,
+    TLSEXT_IDX_record_size_limit,
+    TLSEXT_IDX_application_settings,
     TLSEXT_IDX_padding,
     TLSEXT_IDX_psk,
     /* Dummy index - must always be the last entry */
diff --git a/ssl/statem/extensions.c b/ssl/statem/extensions.c
index 7709393..be589a2 100644
--- a/ssl/statem/extensions.c
+++ b/ssl/statem/extensions.c
@@ -369,6 +369,21 @@ static const EXTENSION_DEFINITION ext_defs[] = {
         tls_construct_certificate_authorities,
         NULL,
     },
+    {
+        TLSEXT_TYPE_status_request_v2,
+        SSL_EXT_CLIENT_HELLO,
+        NULL,
+        NULL, NULL, NULL, NULL, NULL },
+    {
+        TLSEXT_TYPE_record_size_limit,
+        SSL_EXT_CLIENT_HELLO,
+        NULL,
+        NULL, NULL, NULL, NULL, NULL },
+    {
+        TLSEXT_TYPE_application_settings,
+        SSL_EXT_CLIENT_HELLO,
+        NULL,
+        NULL, NULL, NULL, NULL, NULL },
     { /* Must be immediately before pre_shared_key */
         TLSEXT_TYPE_padding,
         SSL_EXT_CLIENT_HELLO,
