diff --git a/src/event/ngx_event_openssl.c b/src/event/ngx_event_openssl.c
index 99ec654..4ce3c05 100644
--- a/src/event/ngx_event_openssl.c
+++ b/src/event/ngx_event_openssl.c
@@ -14,6 +14,7 @@
 #endif
 
 
+#define NGX_SSL_JA_DATA_SIZE 256 /* = (sizeof(uint16_t) * 128) */
 #define NGX_SSL_PASSWORD_BUFFER_SIZE  4096
 
 
@@ -2011,6 +2012,38 @@ ngx_ssl_set_client_hello_callback(ngx_ssl_t *ssl, ngx_ssl_client_hello_arg *cb)
 
 #ifdef SSL_CLIENT_HELLO_SUCCESS
 
+static void
+ngx_ssl_client_hello_ja_cb(SSL *ssl)
+{
+    ngx_connection_t  *c;
+    ngx_str_t         *ja;
+
+    c = ngx_ssl_get_connection(ssl);
+
+    if (c == NULL || c->ssl == NULL) {
+        return;
+    }
+
+    ja = &c->ssl->fp_ja_data;
+    ja->len = NGX_SSL_JA_DATA_SIZE;
+
+    ja->data = ngx_pnalloc(c->pool, NGX_SSL_JA_DATA_SIZE);
+    if (ja->data == NULL) {
+        ja->len = 0;
+        return;
+    }
+
+    ja->len = SSL_client_hello_get_ja_data(ssl, ja->data,
+                                            NGX_SSL_JA_DATA_SIZE);
+    if (ja->len == 0) {
+        ngx_log_error(NGX_LOG_WARN, c->log, 0, "SSL_client_hello_get_ja_data "
+                "seems the buffer size is less that number of fileds; "
+                "for this SSL connection can't get a ja3 hash string");
+        ja->data = NULL;
+    }
+}
+
+
 int
 ngx_ssl_client_hello_callback(ngx_ssl_conn_t *ssl_conn, int *ad, void *arg)
 {
@@ -2025,6 +2058,9 @@ ngx_ssl_client_hello_callback(ngx_ssl_conn_t *ssl_conn, int *ad, void *arg)
     cb = SSL_CTX_get_ex_data(c->ssl->session_ctx,
                              ngx_ssl_client_hello_arg_index);
 
+    /* capture JA fingerprint data before SNI processing */
+    ngx_ssl_client_hello_ja_cb(ssl_conn);
+
     if (SSL_client_hello_get0_ext(ssl_conn, TLSEXT_TYPE_server_name,
                                   (const unsigned char **) &p, &len)
         == 0)
diff --git a/src/event/ngx_event_openssl.h b/src/event/ngx_event_openssl.h
index 79ae395..bc73ef6 100644
--- a/src/event/ngx_event_openssl.h
+++ b/src/event/ngx_event_openssl.h
@@ -149,6 +149,12 @@ struct ngx_ssl_connection_s {
     unsigned                    early_preread:1;
     unsigned                    write_blocked:1;
     unsigned                    sni_accepted:1;
+
+    ngx_str_t                   fp_ja_data;
+    ngx_str_t                   fp_ja3_str;
+    ngx_str_t                   fp_ja3_hash;
+    ngx_str_t                   fp_ja4_str;
+    uint16_t                    fp_tls_greased;
 };
 
 
diff --git a/src/http/v2/ngx_http_v2.c b/src/http/v2/ngx_http_v2.c
index efe2290..2891bc6 100644
--- a/src/http/v2/ngx_http_v2.c
+++ b/src/http/v2/ngx_http_v2.c
@@ -307,6 +307,8 @@ ngx_http_v2_init(ngx_event_t *rev)
         ngx_add_timer(rev, cscf->client_header_timeout);
     }
 
+    h2c->fp_fingerprinted = 0;
+
     c->idle = 1;
     ngx_reusable_connection(c, 0);
 
@@ -1358,6 +1360,14 @@ ngx_http_v2_state_headers(ngx_http_v2_connection_t *h2c, u_char *pos,
         }
     }
 
+    if (!h2c->fp_fingerprinted && h2c->fp_priorities.len + 4 <= NGX_FP_V2_BUFFER_SIZE) {
+        h2c->fp_priorities.data[h2c->fp_priorities.len] = (uint8_t)stream->node->id;
+        h2c->fp_priorities.data[h2c->fp_priorities.len+1] = (uint8_t)excl;
+        h2c->fp_priorities.data[h2c->fp_priorities.len+2] = (uint8_t)depend;
+        h2c->fp_priorities.data[h2c->fp_priorities.len+3] = (uint8_t)(weight-1);
+        h2c->fp_priorities.len += 4;
+    }
+
     return ngx_http_v2_state_header_block(h2c, pos, end);
 
 rst_stream:
@@ -1781,6 +1791,9 @@ ngx_http_v2_state_process_header(ngx_http_v2_connection_t *h2c, u_char *pos,
     }
 
     if (header->name.data[0] == ':') {
+        if (!h2c->fp_fingerprinted && h2c->fp_pseudoheaders.len < NGX_FP_V2_BUFFER_SIZE && header->name.len > 1)
+            h2c->fp_pseudoheaders.data[h2c->fp_pseudoheaders.len++] = header->name.data[1];
+
         rc = ngx_http_v2_pseudo_header(r, header);
 
         if (rc == NGX_OK) {
@@ -2209,6 +2222,12 @@ ngx_http_v2_state_settings_params(ngx_http_v2_connection_t *h2c, u_char *pos,
         ngx_log_debug2(NGX_LOG_DEBUG_HTTP, h2c->connection->log, 0,
                        "http2 setting %ui:%ui", id, value);
 
+        if (!h2c->fp_fingerprinted && h2c->fp_settings.len < NGX_FP_V2_SETTINGS_MAX) {
+            h2c->fp_settings.ids[h2c->fp_settings.len] = (uint16_t)id;
+            h2c->fp_settings.values[h2c->fp_settings.len] = (uint32_t)value;
+            h2c->fp_settings.len++;
+        }
+
         switch (id) {
 
         case NGX_HTTP_V2_INIT_WINDOW_SIZE_SETTING:
@@ -2493,6 +2512,9 @@ ngx_http_v2_state_window_update(ngx_http_v2_connection_t *h2c, u_char *pos,
     }
 
     h2c->send_window += window;
+    if (!h2c->fp_fingerprinted) {
+        h2c->fp_windowupdate = window;
+    }
 
     while (!ngx_queue_empty(&h2c->waiting)) {
         q = ngx_queue_head(&h2c->waiting);
diff --git a/src/http/v2/ngx_http_v2.h b/src/http/v2/ngx_http_v2.h
index 9605c8a..026f3c7 100644
--- a/src/http/v2/ngx_http_v2.h
+++ b/src/http/v2/ngx_http_v2.h
@@ -17,6 +17,9 @@
 
 #define NGX_HTTP_V2_STATE_BUFFER_SIZE    16
 
+#define NGX_FP_V2_BUFFER_SIZE            32
+#define NGX_FP_V2_SETTINGS_MAX           8
+
 #define NGX_HTTP_V2_DEFAULT_FRAME_SIZE   (1 << 14)
 #define NGX_HTTP_V2_MAX_FRAME_SIZE       ((1 << 24) - 1)
 
@@ -121,6 +124,18 @@ typedef struct {
 } ngx_http_v2_hpack_t;
 
 
+typedef struct {
+  u_char data[NGX_FP_V2_BUFFER_SIZE];
+  size_t len;
+} ngx_http_v2_fp_fixed_str_t;
+
+typedef struct {
+  uint16_t   ids[NGX_FP_V2_SETTINGS_MAX];
+  uint32_t   values[NGX_FP_V2_SETTINGS_MAX];
+  ngx_uint_t len;
+} ngx_http_v2_fp_settings_t;
+
+
 struct ngx_http_v2_connection_s {
     ngx_connection_t                *connection;
     ngx_http_connection_t           *http_connection;
@@ -168,6 +183,12 @@ struct ngx_http_v2_connection_s {
     unsigned                         table_update:1;
     unsigned                         blocked:1;
     unsigned                         goaway:1;
+
+    unsigned                         fp_fingerprinted:1;
+    ngx_http_v2_fp_settings_t        fp_settings;
+    ngx_http_v2_fp_fixed_str_t       fp_priorities, fp_pseudoheaders;
+    ngx_uint_t                       fp_windowupdate;
+    ngx_str_t                        fp_str;
 };
 
 
